src/CodeAnalysis/CSharp/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,114 @@
+        private static readonly string s_With = "With";
+
+        private static readonly ImmutableArray<string> s_immutableMethodNames = ImmutableArray.Create(
+            s_Add, 
+            s_Remove, 
+            s_Replace, 
+            s_With);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                var solutionSymbol = compilationContext.Compilation.GetTypeByMetadataName(s_solutionFullName);
+                var projectSymbol = compilationContext.Compilation.GetTypeByMetadataName(s_projectFullName);
+                var documentSymbol = compilationContext.Compilation.GetTypeByMetadataName(s_documentFullName);
+                var syntaxNodeSymbol = compilationContext.Compilation.GetTypeByMetadataName(s_syntaxNodeFullName);
+                var compilationSymbol = compilationContext.Compilation.GetTypeByMetadataName(s_compilationFullName);
+
+                var immutableSymbols = ImmutableArray.Create(solutionSymbol, projectSymbol, documentSymbol, syntaxNodeSymbol, compilationSymbol);
+                //Only register our node action if we can find the symbols for our immutable types
+                if (immutableSymbols.Any(n => n == null))
+                {
+                    return;
+                }
+                compilationContext.RegisterSyntaxNodeAction(sc => AnalyzeInvocationForIgnoredReturnValue(sc, immutableSymbols), SyntaxKind.InvocationExpression);
+            });
+        }
+
+        public void AnalyzeInvocationForIgnoredReturnValue(SyntaxNodeAnalysisContext context, ImmutableArray<INamedTypeSymbol> immutableTypeSymbols)
+        {
+            var model = context.SemanticModel;
+            var candidateInvocation = (InvocationExpressionSyntax)context.Node;
+
+            //We're looking for invocations that are direct children of expression statements
+            if (!(candidateInvocation.Parent is ExpressionStatementSyntax))
+            {
+                return;
+            }
+
+            //If we can't find the method symbol, quit
+            var methodSymbol = model.GetSymbolInfo(candidateInvocation).Symbol as IMethodSymbol;
+            if (methodSymbol == null)
+            {
+                return;
+            }
+
+            //If the method doesn't start with something like "With" or "Replace", quit
+            string methodName = methodSymbol.Name;
+            if (!s_immutableMethodNames.Any(n => methodName.StartsWith(n)))
+            {
+                return;
+            }
+
+            //If we're not in one of the known immutable types, quit
+            var parentType = methodSymbol.ReceiverType as INamedTypeSymbol;
+            if(parentType == null)
+            {
+                return;
+            }
+            
+            var baseTypesAndSelf = methodSymbol.ReceiverType.GetBaseTypes().ToList();
```

Comment:
Is there an API for `IsDerivedFrom()` or something similar that might cache this info as it looks it up? @sharwell expressed concern that `.Add()` is common enough that this might be a performance problem.

If this approach looks fine to you guys, that's alright too.